src/app/public/modules/dynamic-component/dynamic-component.service.ts

```diff
@@ -0,0 +1,80 @@
+//#region imports
+
+import {
+  ApplicationRef,
+  ComponentFactoryResolver,
+  ComponentRef,
+  EmbeddedViewRef,
+  Injectable,
+  Injector,
+  RendererFactory2,
+  Renderer2
+} from '@angular/core';
+
+import {
+  SkyWindowRefService
+} from '../window';
+
+import {
+  SkyDynamicComponentArgs
+} from './dynamic-component-args';
+
+import {
+  SkyDynamicComponentLocation
+} from './dynamic-component-location';
+
+//#endregion
+
+/**
+ * Angular service for creating and rendering a dynamic component.
+ */
+@Injectable()
+export class SkyDynamicComponentService {
+
+  private renderer: Renderer2;
+
+  constructor(
+    private componentFactoryResolver: ComponentFactoryResolver,
+    private appRef: ApplicationRef,
+    private injector: Injector,
+    private windowRef: SkyWindowRefService,
+    rendererFactory: RendererFactory2
+  ) {
+    // Based on suggestions from https://github.com/angular/angular/issues/17824
+    // for accessing an instance of Renderer2 in a service since Renderer2 can't
+    // be injected into a service.  Passing undefined for both parameters results
+    // in the default renderer which is what we want here.
+    this.renderer = rendererFactory.createRenderer(undefined, undefined);
+  }
+
+  /**
+   * Creates an instance of the specified component and adds it to the specified location
+   * on the page.
+   * @param args Options for creating the dynamic component.
+   */
+  public createComponent<T>(args: SkyDynamicComponentArgs): ComponentRef<T> {
+    const cmpRef = this.componentFactoryResolver
+      .resolveComponentFactory<T>(args.componentType)
+      .create(this.injector);
+
+    this.appRef.attachView(cmpRef.hostView);
+
+    // Technique for retrieving the component's root node taken from here:
+    // https://malcoded.com/posts/angular-dynamic-components
+    const el = (cmpRef.hostView as EmbeddedViewRef<any>).rootNodes[0];
+
+    const bodyEl = this.windowRef.getWindow().document.body;
+
+    switch (args.location) {
```

Comment:
Probably not a huge deal, but this option could be confused with the native `window.location` object.

Reply:
Fair enough. I noticed this issue when I attempted to set the object property to a local variable like this (not that we'd ever do this, but it was the reason it came up):

```
const location = args.location;
// Programmer can't type `location` to access the global `location` object
// (they'd have to type `window.location`)
switch (location) {
```

At any rate, if you don't think it's a problem, disregard.